docs/source/scheduler.rst

```diff
@@ -0,0 +1,263 @@
+no capacity pressure on the cell.
+
+Within allocation, applications are ordered by priority. Next step is to
+construct "utilization", which is roughly a ratio of total accumulated capacity
+to reserved. As capacity is a vector (memory, cpu, disk, …) the max of the
+ratio is used.
+
+For each application in the allocation, the scheduler will maintain a tuple of
+(dynamic-rank, utilization-before, utilization-after). Dynamic rank is
+constructed from allocation base rank and rank adjustment attributes. If
+reserved capacity is not exceeded, dynamic rank is base rank minus rank
+adjustment. If reserved capacity is exceeded, dynamic rank equals to base rank.
+
+All private allocation queue are merged, bottom up, sorted by dynamic rank and
+utilization. Allocation structure is hierarchical, and rank, rank adjustment
+and utilization is recalculated on each level.
+
+Top level allocation will contain all applications in the partition, ordered by
+rank and utilization.
+
+At each level, the following invariants are maintained:
+
+ * Applications from allocations with lower rank are first in the queue.
+ * Applications within reserved capacity has rank less or equal to applications
+   above reserved capacity.
+ * Applications within reserved capacity has utilization < 1.
+
+The hierarchical merge allows to achieve the following design goals:
+
+ * Different tenants reserve capacity independently (Treadmill does not allow
+   for reservations to go beyond total capacity of the cell)
+ * Tenant hierarchy allows to offer unused capacity to the sibling first,
+   parents later. Consider someone running grid application, and creating dev
+   and prod sub allocations. If prod is underutilized, capacity will be offered
+   first to the test grid, then to other Treadmill users. Similarly, stopping
+   test grid will boost capacity for prod grid first (sibling).
+ * Favor production applications to test (qa, uat). Typical setup is to
+   configure all allocations with same base rank, but assign different
+   adjustment to allocations for different environments. Prod allocations are
+   configured with rank adjustment 3, uat with 2, qa with 1, dev - 0. As such,
+   global queue will be partitioned as following: prod applications (ordered by
+   utilization) running within reserved capacity, uat applications within
+   reserved capacity, … , dev applications within reserved capacity,
+   non-reserved. Such scheme ensures that in the event of catastrophic capacity
+   failure production application running within reserved capacity will be
+   evicted last.
+
+
+.. _section-topology-constraints:
+
+Server topology constraints
+---------------------------
+
+Servers in the cell are organized in a hierarchy that aims to reflect physical
+topology of the datacenter.
+
+Typical structure is server, rack, pod (rack is collection of servers, pod is
+collection of racks).
+
+Treadmill scheduler supports two placement strategies - stack and spread. The
```

Comment:
I recommend explaining stack and spread placement strategy a bit better. You kind of explain spread, but don't really explain stack.  I found this vague/confusing to read.

Reply:
rewritten